src/execution_plan/execution_plan.c

```diff
@@ -8,695 +8,804 @@
+                 * try to locate and remove it, there's no real harm except some performace hit
+                 * in keeping that label matrix. */
+                if(exp->operands[0].diagonal) AlgebraicExpression_RemoveTerm(exp, 0, NULL);
+                op = NewNodeByLabelScanOp(exp->src_node, record_id);
+            } else {
+                op = NewAllNodeScanOp(gc->g, exp->src_node, record_id);
+            }
 
-    for(int i = 0; i < root->childCount; i++) {
-        ExecutionPlan_Taps(root->children[i], taps);
-    }
-}
+            Vector_Push(traversals, op);
+            _UpdateResolvedVariables(resolved, op);
 
-OpBase* ExecutionPlan_Locate_References(OpBase *root, rax *references) {
-    OpBase *op = NULL;    
-    Vector *temp = _ExecutionPlan_Locate_References(root, &op, references);
-    Vector_Free(temp);
-    return op;
-}
+            for(int i = 0; i < expCount; i++) {
+                exp = exps[i];
+                if(exp->operand_count == 0) continue;
 
-// TODO: get rid of _ReturnClause_GetExpressions, _WithClause_GetExpressions and _OrderClause_GetExpressions.
-/* Returns an array of arithmetic expression, one for every return element.
- * caller is responsible for freeing each arithmetic expression in addition
- * to the array itself. */
-static AR_ExpNode** _ReturnClause_GetExpressions(const AST *ast) {
-    assert(ast->returnNode);
+                uint ast_id;
+                uint src_node_idx;
+                uint dest_node_idx;
+                uint edge_idx = IDENTIFIER_NOT_FOUND;
 
-    AST_ReturnNode *return_node = ast->returnNode;
-    unsigned int elem_count = array_len(return_node->returnElements);
-    AR_ExpNode **exps = array_new(AR_ExpNode*, elem_count);
+                // Make sure that all entities are represented in Record
+                ast_id = exp->src_node->id;
+                src_node_idx = RecordMap_FindOrAddID(segment->record_map, ast_id);
 
-    for(unsigned int i = 0; i < elem_count; i++) {
-        AST_ReturnElementNode *elem = return_node->returnElements[i];
-        AR_ExpNode *exp = AR_EXP_BuildFromAST(ast, elem->exp);
-        exps = array_append(exps, exp);
-    }
+                ast_id = exp->dest_node->id;
+                dest_node_idx = RecordMap_FindOrAddID(segment->record_map, ast_id);
 
-    return exps;
-}
+                if (exp->edge) {
+                    ast_id = exp->edge->id;
+                    edge_idx = RecordMap_FindOrAddID(segment->record_map, ast_id);
+                }
+
+                if(exp->edge && QGEdge_VariableLength(exp->edge)) {
+                    op = NewCondVarLenTraverseOp(exp,
+                            exp->edge->minHops,
+                            exp->edge->maxHops,
+                            src_node_idx,
+                            dest_node_idx,
+                            gc->g);
+                } else {
+                    op = NewCondTraverseOp(gc->g, exp, src_node_idx, dest_node_idx, edge_idx, TraverseRecordCap(ast));
```

Comment:
I'll prefer to keep `NewCondTraverseOp` argument list to a minimum
can we resolve `src_node_idx`, `dest_node_idx`, `edge_idx`, `TraverseRecordCap(ast)` internally?

Reply:
I'd like to keep `TraverseRecordCap` outside so that ops don't have a dependency on the AST, but all the indexes can be resolved internally.

Reply:
Fair enough